presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SymbolMapper.java

```diff
@@ -68,74 +70,108 @@ public Expression rewriteSymbolReference(SymbolReference node, Void context, Exp
+        }
 
-    public TopNNode map(TopNNode node, PlanNode source, PlanNodeId newNodeId)
-    {
-        ImmutableList.Builder<Symbol> symbols = ImmutableList.builder();
-        ImmutableMap.Builder<Symbol, SortOrder> orderings = ImmutableMap.builder();
-        Set<Symbol> seenCanonicals = new HashSet<>(node.getOrderBy().size());
-        for (Symbol symbol : node.getOrderBy()) {
-            Symbol canonical = map(symbol);
-            if (seenCanonicals.add(canonical)) {
-                seenCanonicals.add(canonical);
-                symbols.add(canonical);
-                orderings.put(canonical, node.getOrderings().get(symbol));
+        @Override
+        public PlanNode visitTopN(TopNNode node, RewriteContext<Void> context)
+        {
+            PlanNode source = getOnlyElement(recursionStrategy.process(node, context));
+
+            ImmutableList.Builder<Symbol> symbols = ImmutableList.builder();
+            ImmutableMap.Builder<Symbol, SortOrder> orderings = ImmutableMap.builder();
+            Set<Symbol> seenCanonicals = new HashSet<>(node.getOrderBy().size());
+            for (Symbol symbol : node.getOrderBy()) {
+                Symbol canonical = map(symbol);
+                if (seenCanonicals.add(canonical)) {
+                    seenCanonicals.add(canonical);
+                    symbols.add(canonical);
+                    orderings.put(canonical, node.getOrderings().get(symbol));
+                }
             }
+
+            return new TopNNode(
+                    planNodeIdAllocationStrategy.getId(node),
+                    source,
+                    node.getCount(),
+                    symbols.build(),
+                    orderings.build(),
+                    node.getStep());
         }
 
-        return new TopNNode(
-                newNodeId,
-                source,
-                node.getCount(),
-                symbols.build(),
-                orderings.build(),
-                node.getStep());
-    }
+        @Override
+        public PlanNode visitAggregation(AggregationNode node, RewriteContext<Void> context)
+        {
+            PlanNode source = getOnlyElement(recursionStrategy.process(node, context));
 
-    private List<Symbol> mapAndDistinct(List<Symbol> outputs)
-    {
-        Set<Symbol> added = new HashSet<>();
-        ImmutableList.Builder<Symbol> builder = ImmutableList.builder();
-        for (Symbol symbol : outputs) {
-            Symbol canonical = map(symbol);
-            if (added.add(canonical)) {
-                builder.add(canonical);
+            ImmutableMap.Builder<Symbol, Aggregation> aggregations = ImmutableMap.builder();
+            for (Map.Entry<Symbol, Aggregation> entry : node.getAggregations().entrySet()) {
+                Symbol symbol = entry.getKey();
+                Aggregation aggregation = entry.getValue();
+
+                aggregations.put(map(symbol), new Aggregation(
```

Comment:
I think I do not understand you completely, lets go through the example.

With mapping `('a' -> 'A', 'b' - 'B')` and plan:
```
project ([a: 1, b: 3 + a],  source)
```

So in this case columns `a` and `b` exists in range of mapping, so why should I allocate new symbol and not to reuse just a mapping to produce:
```
project ([A: 1, B: 3 + A],  source)
```
This is how `SymbolMapper` behaves currently. What have I missed?